### PR TITLE
Time-stepping algorithm 

### PR DESCRIPTION
**Pull request by sayerhs**
The time-stepping algorithm discussed during the FY19-Q1 group meeting has been implemented in the Nalu-Wind codebase and tested on several wind problems of interest: airfoil simulations, turbine simulations in uniform inflow. This pull request is to prepare the updates to be merged into the `master` branch. The two main changes introduced by the new time-stepping algorithm are:

- Using the inverse of the diagonal term of the Momentum equation as the projection timescale in pressure equation instead of `\Delta t I`, and 
- An option to use under-relaxtion for the momentum and scalar transport equations (k, omega, and enthalpy, etc.)
- Option to relax the pressure update while using the full pressure correction solution for the projection step

I have tried to make sure that both the original Nalu algorithm and the new time-stepping algorithm can be selected via user input in the input file and tried to preserve the original algorithm as much as possible. However, since the _projection timescale_ is no longer a constant, the term has to appear in the LHS of the continuity equation. While the equations are normalized with `\Delta t`, it still introduces diffs in the regression tests. 

This pull request is not ready for a merge, but is meant for discussions before the final merge. Still to do:

- [x] Re-bless norms for _failing_ regression tests 
- [x] Discuss pathway to HO-kernels to use new time-stepping algorithm
- [x] Update theory manual with descriptions of the new time-stepping algorithm
- [x] Introduce the DU-91 airfoil case as the regression test for the new time-stepping algorithm
- [x] Rebase/squash commits into one logical commit before merge

### CODE REVIEW

**Comment by sayerhs** [comment]
Here is the output for the failing tests

```
elemHybridFluidsShift          1.10229e-03
heliumPlume_rst                4.59092e-04
milestoneRun                   2.37487e-05
variableDensNonUniform         4.30000e-06
dgNonConformal3dFluidsHexTet   2.25218e-06
hoHelium                       8.34363e-07
variableDensNonIso             5.00000e-07
mixedTetPipe                   2.54702e-07
nonIsoNonUniformElemOpenJet    8.88111e-08
heatedWaterChannelEdge         4.32107e-08
heatedWaterChannelElem         4.10979e-08
elemClosedDomain               1.38820e-08
waleElemXflowMixFrac35m        2.59664e-09
dgNonConformal3dFluids         8.46819e-10
kovasznay_P5                   3.21634e-10
nonIsoElemOpenJetConsolidated  1.30616e-10
nonIsoElemOpenJet              1.30616e-10
dgNonConformalThreeBlade       1.20682e-10
ablNeutralEdge                 5.94038e-11
ablUnstableEdge                4.01542e-11
actuatorLine                   3.46583e-11
dgNonConformalElemCylinder     2.57428e-11
ablStableElem                  1.54246e-11
ablUnstableEdge_ra             8.32000e-12
uqSlidingMeshDG                4.34100e-12
movingCylinder                 1.58100e-12
fluidsPmrChtPeriodic           1.39630e-12
dgNonConformalFluids           1.38118e-12
elemBackStepLRSST              8.47700e-13
dgNonConformal3dFluidsP1P2     5.50881e-13
dgNonConformalFluidsEdge       4.19999e-13
ekmanSpiral                    4.41000e-15
ekmanSpiralConsolidated        3.42600e-15
milestoneRunConsolidated       2.99999e-15
oversetFluids                  2.53000e-15
elemPipeCHT                    2.00000e-15
oversetFluidsEdge              1.75000e-15
karmanVortex                   1.59000e-15
```

**Comment by sayerhs** [comment]
@rcknaus Thank you for your comments. I'll rebless the gold files and update documentation. Since you've are actively developing/working on the HO codebase, would you be willing to update the `compute_mdot` and the _variable coefficient Laplacian_ for those kernels?

**Comment by rcknaus** [comment]
Sure, I can do that.  I was also looking a bit more at usual test diffs, and elemHybridFluidsShift seems a bit high.

Test diffs resulting from randomly ordering nodes, averaged over 30 trials.  This was done about a year ago, so it's possible tests are a bit different now.

| name                         | mean diff    | stddev   |
|------------------------------|----------|----------|
| ablNeutralEdge               | 5.74E-11 | 1.40E-11 |
| ablStableElem                | 1.78E-11 | 5.04E-12 |
| ablUnstableEdge              | 3.77E-11 | 8.72E-12 |
| ablUnstableEdge_ra           | 9.50E-12 | 5.56E-12 |
| actuatorLine                 | 1.84E-11 | 6.91E-12 |
| concentricRad                | 5.92E-15 | 3.24E-15 |
| cvfemHC                      | 4.43E-08 | 1.91E-08 |
| dgNonConformal3dFluidsHexTet | 1.71E-09 | 1.23E-09 |
| dgNonConformalElemCylinder   | 6.14E-12 | 1.90E-14 |
| dgNonConformalFluids         | 3.78E-13 | 1.82E-13 |
| dgNonConformalFluidsEdge     | 1.26E-12 | 5.56E-13 |
| dgNonConformalThreeBlade     | 1.16E-10 | 2.71E-11 |
| edgeHybridFluids             | 2.44E-14 | 2.68E-15 |
| ekmanSpiral                  | 3.33E-14 | 4.23E-15 |
| ekmanSpiralConsolidated      | 2.91E-15 | 5.08E-16 |
| elemBackStepLRSST            | 5.95E-13 | 3.39E-13 |
| elemClosedDomain             | 1.21E-08 | 4.65E-09 |
| elemHybridFluids             | 2.29E-15 | 3.06E-16 |
| elemHybridFluidsShift        | 4.17E-12 | 2.50E-12 |
| elemPipeCHT                  | 2.33E-16 | 4.30E-16 |
| fluidsPmrChtPeriodic         | 3.35E-07 | 1.83E-06 |
| heatedWaterChannelEdge       | 4.90E-08 | 2.56E-08 |
| heatedWaterChannelElem       | 1.79E-08 | 8.92E-09 |
| heliumPlume                  | 1.75E-15 | 1.03E-15 |
| heliumPlume_rst              | 6.50E-15 | 1.78E-14 |
| hoVortex                     | 4.69E-17 | 2.57E-16 |
| inputFireEdgeUpwind          | 1.23E-07 | 2.14E-08 |
| inputFireElem                | 2.73E-08 | 2.11E-08 |
| kovasznay_P5                 | 4.11E-10 | 5.78E-10 |
| milestoneRun                 | 2.23E-15 | 1.36E-15 |
| milestoneRunConsolidated     | 1.17E-15 | 1.09E-15 |
| mixedTetPipe                 | 1.50E-15 | 1.24E-15 |
| movingCylinder               | 1.87E-12 | 8.34E-13 |
| overset                      | 2.23E-15 | 9.46E-16 |
| oversetFluids                | 3.38E-15 | 1.68E-15 |
| oversetFluidsEdge            | 2.94E-15 | 1.28E-15 |
| oversetHybrid                | 2.94E-13 | 1.11E-13 |
| steadyTaylorVortex           | 1.04E-16 | 5.72E-16 |
| uqSlidingMeshDG              | 1.05E-12 | 4.22E-13 |
| variableDensNonUniform_P5    | 3.15E-12 | 1.48E-12 |
| waleElemXflowMixFrac3        | 1.54E-14 | 5.04E-15 |

**Comment by sayerhs** [comment]
@rcknaus Yes, I noticed the diffs in `elemHybridFluidsShift` and it is really weird to me that `elemHybridFluids` is just fine but shifted shows this difference. That was the reason to put the code up for you and Ganesh to review with the hope that extra pairs of eyes could spot if anything was amiss.

File: src/kernel/ContinuityAdvElemKernel.C [hunk 1]
```diff
@@ -119,27 +122,34 @@ ContinuityAdvElemKernel<AlgTraits>::execute(
...
     for (int j = 0; j < AlgTraits::nDim_; ++j) {
       w_uIp[j] = 0.0;
       w_rho_uIp[j] = 0.0;
       w_Gpdx_Ip[j] = 0.0;
       w_dpdxIp[j] = 0.0;
     }
 
+    for (int ic = 0; ic < AlgTraits::nodesPerElement_; ++ic) {
+      const DoubleType r = v_shape_function_(ip, ic);
+      projTimeScaleIP += r / v_udiag(ic);
```
Comment (gantech) [inline src/kernel/ContinuityAdvElemKernel.C]: This is really a question to educate myself. Can this go into the loop below? May be GPU's don't like it.

File: src/kernel/MomentumMassElemKernel.C [hunk 2]
```diff
@@ -160,7 +163,7 @@ MomentumMassElemKernel<AlgTraits>::execute(
     for (int ic=0; ic < AlgTraits::nodesPerElement_; ++ic) {
       const int icNdim = ic * AlgTraits::nDim_;
       const DoubleType r = v_shape_function_(ip, ic);
-      const DoubleType lhsfac = r * gamma1_ * rhoNp1 * scV / dt_;
+      const DoubleType lhsfac = r * gamma1_ * rhoNp1 * scV / dt_ * diagRelaxFactor_;
```
Comment (gantech) [inline src/kernel/MomentumMassElemKernel.C]: Is this required? I see that the `diagRelaxFactor_` is 1.0.

Comment (sayerhs) [reply src/kernel/MomentumMassElemKernel.C]: `diagRelaxFactor_` is determined by the relaxation factor provided in the input file by the user. It is defaulted to `1.0` but will not be always `1.0`. See Line 85 in this file.

File: src/kernel/ContinuityAdvElemKernel.C (hunk 1 above)
Comment (sayerhs) [reply src/kernel/ContinuityAdvElemKernel.C]: No, we need a fully computed `projTimeScaleIP` before entering the next loop, it is being used within that loop at line 152

**Pull request merged**